controllers/tc000243_remediation_retry_test.go

```diff
@@ -0,0 +1,225 @@
+apiVersion: infra.contrib.fluxcd.io/v1alpha2
+kind: Terraform
+metadata:
+  name: %s
+  namespace: flux-system
+spec:
+  approvePlan: auto
+  path: ./terraform-hello-world-example
+  remediation:
+    retries: 3
+  retryInterval: 5s
+  sourceRef:
+    kind: GitRepository
+    name: %s
+    namespace: flux-system
+  interval: 10s
+`, terraformName, sourceName)), runnerServer.Scheme)
+	g.Expect(err).ToNot(HaveOccurred())
+
+	It("should be created and attached successfully.")
+	g.Expect(k8sClient.Create(ctx, &helloWorldTF)).Should(Succeed())
+	defer waitResourceToBeDelete(g, &helloWorldTF)
+
+	By("checking that the TF resource existed inside the cluster.")
+	helloWorldTFKey := types.NamespacedName{Namespace: "flux-system", Name: terraformName}
+	createdHelloWorldTF := infrav1.Terraform{}
+	g.Eventually(func() bool {
+		err := k8sClient.Get(ctx, helloWorldTFKey, &createdHelloWorldTF)
+		return err == nil
+	}, timeout, interval).Should(BeTrue())
+
+	It("should be reconciled and contain some status conditions.")
+	By("checking that the TF resource's status conditions has 1 element.")
+	g.Eventually(func() int {
+		err := k8sClient.Get(ctx, helloWorldTFKey, &createdHelloWorldTF)
+		if err != nil {
+			return -1
+		}
+		return len(createdHelloWorldTF.Status.Conditions)
+	}, timeout, interval).Should(Equal(1))
+
+	It("should stop retry")
+	By("when reached retry limit specified in .Spec.Remediation.Retries")
+	type failedStatusCheckResult struct {
+		Type    string
+		Reason  string
+		Message string
+		Status  metav1.ConditionStatus
+		Retries int64
+	}
+	expected := failedStatusCheckResult{
+		Type:    meta.StalledCondition,
+		Reason:  infrav1.RetryLimitReachedReason,
+		Message: "Resource reached maximum number of retries.",
+		Status:  metav1.ConditionTrue,
+	}
+	g.Eventually(func() interface{} {
+		err := k8sClient.Get(ctx, helloWorldTFKey, &createdHelloWorldTF)
+		if err != nil {
+			return nil
+		}
+		for _, c := range createdHelloWorldTF.Status.Conditions {
+			if c.Type == expected.Type {
+				return failedStatusCheckResult{
+					Type:    c.Type,
+					Reason:  c.Reason,
+					Message: c.Message,
+					Status:  c.Status,
+				}
+			}
+		}
+		return createdHelloWorldTF.Status
+	}, timeout, interval).Should(Equal(expected))
+	g.Expect(createdHelloWorldTF.Status.ReconciliationFailures).To(Equal(int64(3)))
+
+	reconcileResult, err := reconciler.Reconcile(ctx, controllerruntime.Request{
+		NamespacedName: types.NamespacedName{
+			Namespace: createdHelloWorldTF.GetNamespace(),
+			Name:      createdHelloWorldTF.GetName(),
+		},
+	})
+	g.Expect(err).To(Succeed())
+	g.Expect(reconcileResult.Requeue).To(BeFalse())
+
+	recheckHelloWorldTF := infrav1.Terraform{}
+	g.Eventually(func() interface{} {
+		return k8sClient.Get(ctx, helloWorldTFKey, &recheckHelloWorldTF)
+	}, timeout, interval).Should(Succeed())
+
+	var retryCondition *metav1.Condition
+	for _, cond := range recheckHelloWorldTF.Status.Conditions {
```

Comment:
[`apimeta.FindStatusCondition`](https://pkg.go.dev/k8s.io/apimachinery/pkg/api/meta#FindStatusCondition) will do this bit for you (and you would have to check that it's stalled for the right reason).